baseplate.go

```diff
@@ -0,0 +1,245 @@
+type Baseplate interface {
+	io.Closer
+
+	Config() Config
+	EdgeContextImpl() *edgecontext.Impl
+	Secrets() *secrets.Store
+}
+
+// Server is the primary interface for baseplate servers.
+type Server interface {
+	// Baseplate returns the Baseplate object the server is built on.
+	Baseplate() Baseplate
+
+	// Serve should start the Server on the Addr given by the Config and only
+	// return once the Server has stopped.
+	//
+	// It is recommended that you use baseplate.Serve() rather than calling Serve
+	// directly as baseplate.Serve will manage starting your service as well as
+	// shutting it down gracefully.
+	Serve() error
+
+	// Stop should stop the server gracefully and only return after the server has
+	// finished shutting down.
+	//
+	// It is recommended that you use baseplate.Serve() rather than calling Stop
+	// directly as baseplate.Serve will manage starting your service as well as
+	// shutting it down gracefully in response to a shutdown signal.
+	Stop() error
+}
+
+// Serve runs the given Server until it is given an external shutdown signal
+// using runtimebp.HandleShutdown to handle the signal and shut down the
+// server gracefully.
+//
+// If a StopTimeout is configure, Serve will wait for that duration for the
+// server to stop before timing out and returning to force a shutdown.
+//
+// This is the recommended way to run a Baseplate Server rather than calling
+// server.Start/Stop directly.
+func Serve(ctx context.Context, server Server) {
+	go runtimebp.HandleShutdown(
+		ctx,
+		func(signal os.Signal) {
+			timeout := server.Baseplate().Config().StopTimeout
+			ctx := context.Background()
+			if timeout != 0 {
+				var cancel context.CancelFunc
+				ctx, cancel = context.WithTimeout(ctx, timeout)
+				defer cancel()
+			}
+
+			var err error
+			ch := make(chan error)
+			go func() {
+				ch <- server.Stop()
+			}()
+
+			// Wait for either the context to timeout or Stop() to finish.
+			select {
+			case <-ctx.Done():
+				err = ctx.Err()
+			case e := <-ch:
+				err = e
+			}
+
+			log.Infow(
+				"graceful shutdown",
+				"signal", signal,
+				"stop error", err,
+			)
+		},
+	)
+	log.Info(server.Serve())
+}
+
+func parseConfig(path string) (Config, error) {
+	cfg := Config{}
+	if path == "" {
+		return cfg, errors.New("baseplate.parseConfig: no config path given")
+	}
+
+	f, err := os.Open(path)
+	if err != nil {
+		return cfg, err
+	}
+	defer f.Close()
+
+	if err = yaml.NewDecoder(f).Decode(&cfg); err != nil {
+		return cfg, err
+	}
+
+	log.Debugf("%#v", cfg)
+	return cfg, err
+}
+
+type cancelCloser struct {
+	cancel context.CancelFunc
+}
+
+func (c cancelCloser) Close() error {
+	c.cancel()
+	return nil
+}
+
+// New parses the config file at the given path, initializes the monitoring and
+// logging frameworks, and returns the "serve" context and a new Baseplate to
+// run your service on.
+func New(path string) (context.Context, Baseplate, error) {
```

Comment:
I think this should still accept a `context.Context` rather than producing its own, for instance for testing. I think `context.Context` is so idiomatic in Go we don't need to necessarily lift that part of the shoulders of baseplate users?